File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.rb
# @param {Integer[]} nums
# @return {Integer}
=begin

nums = [3, 4, 5, 1, 2]
        l     m      r
compare with right element
    if m > r:
        move l to m + 1
    else
        move r to m
[1, 2]

[2, 1]
=end
def find_min(nums)
    l = 0
    r = nums.length - 1
    
    while l < r do
    
        mid = (l + r) / 2
        
        if nums[mid] > nums[r]
            l = mid + 1
        else
            r = mid
        end
    end
    nums[l]
    
end

File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.py
"""
                    l        
                 m        
                    r
nums = [4, 5, 6, 7, 0, 1, 2]
                    ^


"""
class Solution:
    def findMin(self, nums: List[int]) -> int:
        
        l, r = 0, len(nums) - 1
        
        while l < r:
            mid = (l + r) // 2
            
            if nums[mid] > nums[r]:
                l = mid + 1
            else:
                r = mid
        
            # print(l)
        return nums[l]
    
"""
          l
          m
             r
[3, 4, 5, 1, 2]

 0  1  2  3  4

"""